--- a/zstd.c
+++ b/zstd.c
@@ -160,7 +160,8 @@
 #define ZSTD_DEPS_IO
 
 #include <stdio.h>
-#define ZSTD_DEBUG_PRINT(...) fprintf(stderr, __VA_ARGS__)
+//#define ZSTD_DEBUG_PRINT(...) fprintf(stderr, __VA_ARGS__)
+#define ZSTD_DEBUG_PRINT(...) do { } while (0)  /* TC: no-op macro remove stderr reference found during R CMD check */
 
 #endif /* ZSTD_DEPS_IO */
 #endif /* ZSTD_DEPS_NEED_IO */
@@ -728,14 +729,14 @@
 #endif
 
 /* disable warnings */
-#ifdef _MSC_VER    /* Visual Studio */
-#  include <intrin.h>                    /* For Visual 2005 */
-#  pragma warning(disable : 4100)        /* disable: C4100: unreferenced formal parameter */
-#  pragma warning(disable : 4127)        /* disable: C4127: conditional expression is constant */
-#  pragma warning(disable : 4204)        /* disable: C4204: non-constant aggregate initializer */
-#  pragma warning(disable : 4214)        /* disable: C4214: non-int bitfields */
-#  pragma warning(disable : 4324)        /* disable: C4324: padded structure */
-#endif
+// #ifdef _MSC_VER    /* Visual Studio */
+// #  include <intrin.h>                    /* For Visual 2005 */
+// #  pragma warning(disable : 4100)        /* disable: C4100: unreferenced formal parameter */
+// #  pragma warning(disable : 4127)        /* disable: C4127: conditional expression is constant */
+// #  pragma warning(disable : 4204)        /* disable: C4204: non-constant aggregate initializer */
+// #  pragma warning(disable : 4214)        /* disable: C4214: non-int bitfields */
+// #  pragma warning(disable : 4324)        /* disable: C4324: padded structure */
+// #endif
 
 /* compile time determination of SIMD support */
 #if !defined(ZSTD_NO_INTRINSICS)
@@ -7619,9 +7620,9 @@
 /**** ended inlining pool.h ****/
 
 /* ======   Compiler specifics   ====== */
-#if defined(_MSC_VER)
-#  pragma warning(disable : 4204)        /* disable: C4204: non-constant aggregate initializer */
-#endif
+// #if defined(_MSC_VER)
+// #  pragma warning(disable : 4204)        /* disable: C4204: non-constant aggregate initializer */
+// #endif
 
 
 #ifdef ZSTD_MULTITHREAD
@@ -10646,9 +10647,9 @@
 /* *************************************
 *  Compiler Specific Options
 ***************************************/
-#ifdef _MSC_VER /* Visual Studio warning fix */
-#  pragma warning(disable : 4127) /* disable: C4127: conditional expression is constant */
-#endif
+// #ifdef _MSC_VER /* Visual Studio warning fix */
+// #  pragma warning(disable : 4127) /* disable: C4127: conditional expression is constant */
+// #endif
 
 #if XXH_NO_INLINE_HINTS  /* disable inlining hints */
 #  if defined(__GNUC__) || defined(__clang__)
@@ -16772,9 +16773,9 @@
 /* **************************************************************
 *  Compiler specifics
 ****************************************************************/
-#ifdef _MSC_VER    /* Visual Studio */
-#  pragma warning(disable : 4127)        /* disable: C4127: conditional expression is constant */
-#endif
+// #ifdef _MSC_VER    /* Visual Studio */
+// #  pragma warning(disable : 4127)        /* disable: C4127: conditional expression is constant */
+// #endif
 
 
 /* **************************************************************
@@ -37349,9 +37350,9 @@
 
 
 /* ======   Compiler specifics   ====== */
-#if defined(_MSC_VER)
-#  pragma warning(disable : 4204)   /* disable: C4204: non-constant aggregate initializer */
-#endif
+// #if defined(_MSC_VER)
+// #  pragma warning(disable : 4204)   /* disable: C4204: non-constant aggregate initializer */
+// #endif
 
 
 /* ======   Dependencies   ====== */
@@ -47144,11 +47145,12 @@
 static int g_displayLevel = 0;
 #endif
 #undef  DISPLAY
-#define DISPLAY(...)                                                           \
+/*#define DISPLAY(...)                                                           \
   {                                                                            \
     fprintf(stderr, __VA_ARGS__);                                              \
     fflush(stderr);                                                            \
-  }
+  }*/
+#define DISPLAY(...) do { } while (0)  /* TC: no-op macro remove stderr reference found during R CMD check */
 #undef  LOCALDISPLAYLEVEL
 #define LOCALDISPLAYLEVEL(displayLevel, l, ...)                                \
   if (displayLevel >= l) {                                                     \
@@ -47686,7 +47688,7 @@
   const unsigned nbTrainSamples = splitPoint < 1.0 ? (unsigned)((double)nbSamples * splitPoint) : nbSamples;
   const unsigned nbTestSamples = splitPoint < 1.0 ? nbSamples - nbTrainSamples : nbSamples;
   const size_t trainingSamplesSize = splitPoint < 1.0 ? COVER_sum(samplesSizes, nbTrainSamples) : totalSamplesSize;
-  const size_t testSamplesSize = splitPoint < 1.0 ? COVER_sum(samplesSizes + nbTrainSamples, nbTestSamples) : totalSamplesSize;
+  //const size_t testSamplesSize = splitPoint < 1.0 ? COVER_sum(samplesSizes + nbTrainSamples, nbTestSamples) : totalSamplesSize; /* TC: unused once DISPLAY macros became no-op */
   /* Checks */
   if (totalSamplesSize < MAX(d, sizeof(U64)) ||
       totalSamplesSize >= (size_t)COVER_MAX_SAMPLES_SIZE) {
@@ -47708,8 +47710,8 @@
   memset(ctx, 0, sizeof(*ctx));
   DISPLAYLEVEL(2, "Training on %u samples of total size %u\n", nbTrainSamples,
                (unsigned)trainingSamplesSize);
-  DISPLAYLEVEL(2, "Testing on %u samples of total size %u\n", nbTestSamples,
-               (unsigned)testSamplesSize);
+  //DISPLAYLEVEL(2, "Testing on %u samples of total size %u\n", nbTestSamples,
+  //             (unsigned)testSamplesSize); /* TC: unused once DISPLAY macros became no-op */
   ctx->samples = samples;
   ctx->samplesSizes = samplesSizes;
   ctx->nbSamples = nbSamples;
@@ -48256,12 +48258,12 @@
   const unsigned kMaxK = parameters->k == 0 ? 2000 : parameters->k;
   const unsigned kSteps = parameters->steps == 0 ? 40 : parameters->steps;
   const unsigned kStepSize = MAX((kMaxK - kMinK) / kSteps, 1);
-  const unsigned kIterations =
-      (1 + (kMaxD - kMinD) / 2) * (1 + (kMaxK - kMinK) / kStepSize);
+  //const unsigned kIterations =
+  //    (1 + (kMaxD - kMinD) / 2) * (1 + (kMaxK - kMinK) / kStepSize); /* TC: unused once DISPLAY macros became no-op */
   const unsigned shrinkDict = 0;
   /* Local variables */
   const int displayLevel = parameters->zParams.notificationLevel;
-  unsigned iteration = 1;
+  //unsigned iteration = 1; /* TC: unused once DISPLAY macros became no-op */
   unsigned d;
   unsigned k;
   COVER_best_t best;
@@ -48297,8 +48299,8 @@
   /* Turn down global display level to clean up display at level 2 and below */
   g_displayLevel = displayLevel == 0 ? 0 : displayLevel - 1;
   /* Loop through d first because each new value needs a new context */
-  LOCALDISPLAYLEVEL(displayLevel, 2, "Trying %u different sets of parameters\n",
-                    kIterations);
+  //LOCALDISPLAYLEVEL(displayLevel, 2, "Trying %u different sets of parameters\n",
+  //                  kIterations); /* TC: unused once DISPLAY macros became no-op */
   for (d = kMinD; d <= kMaxD; d += 2) {
     /* Initialize the context for this value of d */
     COVER_ctx_t ctx;
@@ -48353,9 +48355,9 @@
         COVER_tryParameters(data);
       }
       /* Print status */
-      LOCALDISPLAYUPDATE(displayLevel, 2, "\r%u%%       ",
-                         (unsigned)((iteration * 100) / kIterations));
-      ++iteration;
+      //LOCALDISPLAYUPDATE(displayLevel, 2, "\r%u%%       ",
+      //                   (unsigned)((iteration * 100) / kIterations));
+      //++iteration; /* TC: unused once DISPLAY macros became no-op */
     }
     COVER_best_wait(&best);
     COVER_ctx_destroy(&ctx);
@@ -48406,14 +48408,14 @@
  */
 
 /*- Compiler specifics -*/
-#ifdef __clang__
-#pragma clang diagnostic ignored "-Wshorten-64-to-32"
-#endif
-
-#if defined(_MSC_VER)
-#  pragma warning(disable : 4244)
-#  pragma warning(disable : 4127)    /* C4127 : Condition expression is constant */
-#endif
+// #ifdef __clang__
+// #pragma clang diagnostic ignored "-Wshorten-64-to-32"
+// #endif
+
+// #if defined(_MSC_VER)
+// #  pragma warning(disable : 4244)
+// #  pragma warning(disable : 4127)    /* C4127 : Condition expression is constant */
+// #endif
 
 
 /*- Dependencies -*/
@@ -50408,11 +50410,12 @@
 static int g_displayLevel = 0;
 #endif
 #undef  DISPLAY
-#define DISPLAY(...)                                                           \
+/*#define DISPLAY(...)                                                           \
   {                                                                            \
     fprintf(stderr, __VA_ARGS__);                                              \
     fflush(stderr);                                                            \
-  }
+  }*/
+#define DISPLAY(...) do { } while (0)  /* TC: no-op macro remove stderr reference found during R CMD check */
 #undef  LOCALDISPLAYLEVEL
 #define LOCALDISPLAYLEVEL(displayLevel, l, ...)                                \
   if (displayLevel >= l) {                                                     \
@@ -50676,7 +50679,7 @@
     const unsigned nbTrainSamples = splitPoint < 1.0 ? (unsigned)((double)nbSamples * splitPoint) : nbSamples;
     const unsigned nbTestSamples = splitPoint < 1.0 ? nbSamples - nbTrainSamples : nbSamples;
     const size_t trainingSamplesSize = splitPoint < 1.0 ? COVER_sum(samplesSizes, nbTrainSamples) : totalSamplesSize;
-    const size_t testSamplesSize = splitPoint < 1.0 ? COVER_sum(samplesSizes + nbTrainSamples, nbTestSamples) : totalSamplesSize;
+    //const size_t testSamplesSize = splitPoint < 1.0 ? COVER_sum(samplesSizes + nbTrainSamples, nbTestSamples) : totalSamplesSize; /* TC: unused once DISPLAY macros became no-op */
 
     /* Checks */
     if (totalSamplesSize < MAX(d, sizeof(U64)) ||
@@ -50702,8 +50705,8 @@
     memset(ctx, 0, sizeof(*ctx));
     DISPLAYLEVEL(2, "Training on %u samples of total size %u\n", nbTrainSamples,
                     (unsigned)trainingSamplesSize);
-    DISPLAYLEVEL(2, "Testing on %u samples of total size %u\n", nbTestSamples,
-                    (unsigned)testSamplesSize);
+        //DISPLAYLEVEL(2, "Testing on %u samples of total size %u\n", nbTestSamples,
+        //            (unsigned)testSamplesSize); /* TC: unused once DISPLAY macros became no-op */
 
     ctx->samples = samples;
     ctx->samplesSizes = samplesSizes;
@@ -50987,14 +50990,14 @@
     const unsigned kMaxK = parameters->k == 0 ? 2000 : parameters->k;
     const unsigned kSteps = parameters->steps == 0 ? 40 : parameters->steps;
     const unsigned kStepSize = MAX((kMaxK - kMinK) / kSteps, 1);
-    const unsigned kIterations =
-        (1 + (kMaxD - kMinD) / 2) * (1 + (kMaxK - kMinK) / kStepSize);
+    //const unsigned kIterations =
+    //    (1 + (kMaxD - kMinD) / 2) * (1 + (kMaxK - kMinK) / kStepSize); /* TC: unused once DISPLAY macros became no-op */
     const unsigned f = parameters->f == 0 ? DEFAULT_F : parameters->f;
     const unsigned accel = parameters->accel == 0 ? DEFAULT_ACCEL : parameters->accel;
     const unsigned shrinkDict = 0;
     /* Local variables */
     const int displayLevel = (int)parameters->zParams.notificationLevel;
-    unsigned iteration = 1;
+    //unsigned iteration = 1; /* TC: unused once DISPLAY macros became no-op */
     unsigned d;
     unsigned k;
     COVER_best_t best;
@@ -51036,8 +51039,8 @@
     /* Turn down global display level to clean up display at level 2 and below */
     g_displayLevel = displayLevel == 0 ? 0 : displayLevel - 1;
     /* Loop through d first because each new value needs a new context */
-    LOCALDISPLAYLEVEL(displayLevel, 2, "Trying %u different sets of parameters\n",
-                      kIterations);
+    //LOCALDISPLAYLEVEL(displayLevel, 2, "Trying %u different sets of parameters\n",
+    //                  kIterations); /* TC: unused once DISPLAY macros became no-op */
     for (d = kMinD; d <= kMaxD; d += 2) {
       /* Initialize the context for this value of d */
       FASTCOVER_ctx_t ctx;
@@ -51093,9 +51096,9 @@
           FASTCOVER_tryParameters(data);
         }
         /* Print status */
-        LOCALDISPLAYUPDATE(displayLevel, 2, "\r%u%%       ",
-                           (unsigned)((iteration * 100) / kIterations));
-        ++iteration;
+      //LOCALDISPLAYUPDATE(displayLevel, 2, "\r%u%%       ",
+      //                   (unsigned)((iteration * 100) / kIterations));
+      //++iteration; /* TC: unused once DISPLAY macros became no-op */
       }
       COVER_best_wait(&best);
       FASTCOVER_ctx_destroy(&ctx);
@@ -51143,6 +51146,12 @@
 *  Compiler Options
 ****************************************/
 /* Unix Large Files support (>4GB) */
+
+#ifdef _FILE_OFFSET_BITS /* TC: this macro was already defined on rhub solaris as 32, but needs to be 64 */
+#  undef _FILE_OFFSET_BITS
+#endif
+#define _FILE_OFFSET_BITS 64
+
 #define _FILE_OFFSET_BITS 64
 #if (defined(__sun__) && (!defined(__LP64__)))   /* Sun Solaris 32-bits requires specific definitions */
 #  ifndef _LARGEFILE_SOURCE
@@ -51196,11 +51205,13 @@
 *  Console display
 ***************************************/
 #undef  DISPLAY
-#define DISPLAY(...)         do { fprintf(stderr, __VA_ARGS__); fflush( stderr ); } while (0)
+/*#define DISPLAY(...)         do { fprintf(stderr, __VA_ARGS__); fflush( stderr ); } while (0)*/
+#define DISPLAY(...) do { } while (0)  /* TC: no-op macro remove stderr reference found during R CMD check */
 #undef  DISPLAYLEVEL
-#define DISPLAYLEVEL(l, ...) do { if (notificationLevel>=l) { DISPLAY(__VA_ARGS__); } } while (0)    /* 0 : no display;   1: errors;   2: default;  3: details;  4: debug */
+/*#define DISPLAYLEVEL(l, ...) do { if (notificationLevel>=l) { DISPLAY(__VA_ARGS__); } } while (0)*/   /* 0 : no display;   1: errors;   2: default;  3: details;  4: debug */
+#define DISPLAYLEVEL(l, ...) do { } while (0)  /* TC: no-op macro remove stderr reference found during R CMD check */
 
-static clock_t ZDICT_clockSpan(clock_t nPrevious) { return clock() - nPrevious; }
+//static clock_t ZDICT_clockSpan(clock_t nPrevious) { return clock() - nPrevious; } /* TC: unused once display macros became no-op */
 
 static void ZDICT_printHex(const void* ptr, size_t length)
 {
@@ -51595,11 +51606,11 @@
     BYTE* doneMarks = (BYTE*)malloc((bufferSize+16)*sizeof(*doneMarks));   /* +16 for overflow security */
     U32* filePos = (U32*)malloc(nbFiles * sizeof(*filePos));
     size_t result = 0;
-    clock_t displayClock = 0;
-    clock_t const refreshRate = CLOCKS_PER_SEC * 3 / 10;
+    //clock_t displayClock = 0;  /* TC: unused once display macros became no-op */
+    //clock_t const refreshRate = CLOCKS_PER_SEC * 3 / 10;  /* TC: unused once display macros became no-op */
 
 #   undef  DISPLAYUPDATE
-#   define DISPLAYUPDATE(l, ...)                                   \
+/*#   define DISPLAYUPDATE(l, ...)                                   \
         do {                                                       \
             if (notificationLevel>=l) {                            \
                 if (ZDICT_clockSpan(displayClock) > refreshRate) { \
@@ -51608,7 +51619,8 @@
                 }                                                  \
                 if (notificationLevel>=4) fflush(stderr);          \
             }                                                      \
-        } while (0)
+        } while (0)*/
+#   define DISPLAYUPDATE(l, ...) do { } while (0)  /* TC: no-op macro remove stderr reference found during R CMD check */
 
     /* init */
     DISPLAYLEVEL(2, "\r%70s\r", "");   /* clean display line */
@@ -52128,7 +52140,7 @@
     /* display best matches */
     if (params.zParams.notificationLevel>= 3) {
         unsigned const nb = MIN(25, dictList[0].pos);
-        unsigned const dictContentSize = ZDICT_dictSize(dictList);
+        //unsigned const dictContentSize = ZDICT_dictSize(dictList);  /* TC: unused once display macros became no-op */
         unsigned u;
         DISPLAYLEVEL(3, "\n %u segments found, of total size %u \n", (unsigned)dictList[0].pos-1, dictContentSize);
         DISPLAYLEVEL(3, "list %u best segments \n", nb-1);
